onnxruntime/core/providers/cpu/activation/activations.h

```diff
@@ -17,62 +19,108 @@ namespace onnxruntime {
 template <typename T>
 class Elu final : public OpKernel {
  public:
-  Elu(const OpKernelInfo& info) : OpKernel(info), alpha_(info.GetAttrOrDefault("alpha", 1.0f)) {}
+  Elu(const OpKernelInfo& info) : OpKernel(info) {
+    ORT_THROW_IF_ERROR(info.GetAttr("alpha", &alpha_));
+  }
 
   Status Compute(OpKernelContext* context) const override {
-    const auto* X = context->Input<Tensor>(0);
+    const Tensor* X = context->Input<Tensor>(0);
     Tensor* Y = context->Output(0, X->Shape());
-    EIGEN_X_VAR(xm);
-    EIGEN_Y = (xm >= 0).select(xm, (T)alpha_ * (xm.exp() - 1));
+    float alpha = alpha_;
+    ThreadPool* tp = context->GetOperatorThreadPool();
+    const int64_t input_size = X->Shape().Size();
+    std::ptrdiff_t batch_size = static_cast<std::ptrdiff_t>(input_size);
+    // The cost comes from microbenchmark(manual tunning).
+    const double cost = 40.0;
+    const T* data = X->template Data<T>();
+    T* output = Y->template MutableData<T>();
+    ThreadPool::TryParallelFor(tp, batch_size, cost, [alpha, data, output](ptrdiff_t first, ptrdiff_t last) {
```

Comment:
The pattern here seems similar to FastGelu. Should we use that instead?

Reply:
That one is better for a specific scenario and a specific customer. It's their preference, so I stopped arguing on that. But it doesn't mean we should apply the same pattern on the other code. For example, when you put 4096 there, did you run the perf test by your self? Is it the best number to have? Should it be different for different ops?

Reply:
Yeah, it's difficult to argue for model specific behavior.

Reply:
The pitfall of TryBatchParallelFor is it uses a relative large block size for each thread pool task, sometimes it is more efficient because there were less tasks and less scheduling cost, however, it may also cause inefficient cpu usage due to load balancing issues and stragglers. The stragglers may not be seen if you run the perf test in an ideal environment. For example, all the input data are evenly sample,  the cpu time of each sub task is strictly linear to input size(last - first), and no background task interferes the execution. But it's not the real case. So the code performs a little better(5%-10%) in the ideal environment doesn't mean it would be better in a broader usage.

Reply:
I think it is good to use cost based approach for general activation ops. Here we assume that cost based approach is fully tuned (Need some study to see its advantage than other naive approach as used in FastGelu).

Using cost will have similar problem like the chunk size (4096) parameter in FastGelu: whether the parameter tuned in your machine and input sizes could fit a model with different input size in another machine.

Sometime, an operator level perf test might not be ideal. For example, give an input size of 4000 on 4 cores. If you evenly split it into 4 threads, it might be best for it. If you split it to 2 threads, although it might not be optimal for this op, but it saves two cores that could use in other place (like inter-op) and might gain on model level.

Another thing that can be re-visited is the usage of Eigen vector. Maybe need a benchmark to see the benefit of it on  binary with/without openmp.

Reply:
About the tuning: we don't need to get the best estimation, currently we only need to make sure the new implementation is better than old one whatever the input size is. So it's often I choose a lower than real value, to avoid perf regression(comparing to single thread), but the downside is the perf isn't optimal.  For example, in Gelu, the current implementation indeed has worse performance than single thread when input size is small(50% worse). But it can have better(10%-20%) perf than the cost based version written by me.  So we must have a choice which we want. In the other ops, I prefer to get a better implementation first, as long as it is better than the master. If later one someone make it better, please go ahead.